dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/common/hibernate/InternalHibernateGenericStoreImpl.java

```diff
@@ -268,66 +293,81 @@ public List<Function<Root<T>, Predicate>> getDataSharingPredicates(
   }
 
-  /**
-   * @deprecated use {@link #sharingEnabled( CurrentUserDetails )} instead.
-   */
-  @Deprecated
-  protected boolean sharingEnabled(User user) {
-    boolean b = forceAcl();
-
-    if (b) {
-      return b;
-    } else {
-      return (aclService.isClassShareable(clazz) && !(user == null || user.isSuper()));
-    }
-  }
-
-  protected boolean sharingEnabled(CurrentUserDetails user) {
+  protected boolean sharingEnabled(UserDetails userDetails) {
     boolean b = forceAcl();
 
     if (b) {
       return b;
     } else {
-      return (aclService.isClassShareable(clazz) && !(user == null || user.isSuper()));
+      return (aclService.isClassShareable(clazz)
+          && !(userDetails == null || userDetails.isSuper()));
     }
   }
 
-  protected boolean dataSharingEnabled(CurrentUserDetails user) {
-    return aclService.isDataClassShareable(clazz) && !user.isSuper();
-  }
-
-  /**
-   * @deprecated use {@link #dataSharingEnabled( CurrentUserDetails )} instead.
-   */
-  @Deprecated
-  private boolean dataSharingEnabled(User user) {
-    return aclService.isDataClassShareable(clazz) && !user.isSuper();
+  protected boolean dataSharingDisabled(UserDetails userDetails) {
+    return !aclService.isDataClassShareable(clazz) || userDetails.isSuper();
   }
 
   private List<Function<Root<T>, Predicate>> getSharingPredicates(
-      CriteriaBuilder builder, User user, CurrentUserGroupInfo groupInfo, String access) {
-    if (user == null || groupInfo == null || !sharingEnabled(user)) {
+      CriteriaBuilder builder,
+      UserDetails userDetails,
+      CurrentUserGroupInfo groupInfo,
+      String access) {
+
+    if (userDetails == null || groupInfo == null || !sharingEnabled(userDetails)) {
       return List.of();
     }
 
     return getSharingPredicates(
         builder, groupInfo.getUserUID(), groupInfo.getUserGroupUIDs(), access);
   }
+
+  @Override
+  // TODO: MAS can this be removed and we rely on first fetch on login? make sure current logged in
+  // users are get invalidated when group changes
+  public CurrentUserGroupInfo getCurrentUserGroupInfo(String userUID) {
+    CriteriaBuilder builder = getCriteriaBuilder();
+    CriteriaQuery<Object[]> query = builder.createQuery(Object[].class);
+    Root<User> root = query.from(User.class);
+    query.where(builder.equal(root.get("uid"), userUID));
+    query.select(builder.array(root.get("uid"), root.join("groups", JoinType.LEFT).get("uid")));
+
+    Session session = getSession();
+    List<Object[]> results = session.createQuery(query).getResultList();
+
+    CurrentUserGroupInfo currentUserGroupInfo = new CurrentUserGroupInfo();
+
+    if (CollectionUtils.isEmpty(results)) {
+      currentUserGroupInfo.setUserUID(userUID);
+      return currentUserGroupInfo;
+    }
+
+    for (Object[] result : results) {
+      if (currentUserGroupInfo.getUserUID() == null) {
+        currentUserGroupInfo.setUserUID(result[0].toString());
```

Comment:
is it possible that result[0] could be null?

Reply:
Don't think so, this code is been unmodified in the codebase for a long time.